drivers: crypto: stm32: h7: add AES CCM and GCM support

GeorgeCGV opened the pull request:
Adds support for AES-CCM and AES-GCM cipher modes.

The support is limited to the following STM32 H7 SoCs:
`STM32H723XX`, `STM32H725XX`, `STM32H730XX`, `STM32H730XXQ`, and `STM32H735XX`, as they share the same reference manual.

Note: testing was performed on the `STM32H730XX`.

Crypto sample output:
```yml
Cipher Sample
ECB Mode
Output length (encryption): 16
ECB mode ENCRYPT - Match
Output length (decryption): 16
ECB mode DECRYPT - Match
CBC Mode
Output length (encryption): 80
CBC mode ENCRYPT - Match
Output length (decryption): 64
CBC mode DECRYPT - Match
CTR Mode
Output length (encryption): 64
CTR mode ENCRYPT - Match
Output length (decryption): 64
CTR mode DECRYPT - Match
CCM Mode
Output length (encryption): 23
CCM mode ENCRYPT - Match
Output length (decryption): 23
CCM mode DECRYPT - Match
GCM Mode
Output length (encryption): 42
GCM mode ENCRYPT - Match
Output length (decryption): 42
GCM mode DECRYPT - Match
```

The pull request is open for review but not yet ready to be merged (_therefore, labeled `DNM`_) due to an issue in the STM32 H7 HAL related to GCM tag generation during decryption: https://github.com/STMicroelectronics/stm32h7xx-hal-driver/issues/88.

GeorgeCGV committed:
drivers: crypto: stm32: move device_sem from do_aes

Device semaphore handling is being moved out of the do_aes function.
This is necessary to support upcoming GCM and CCM modes,
which require the semaphore to be held during tag generation.

Signed-off-by: Georgij Černyšiov <geo.cgv@gmail.com>

GeorgeCGV committed:
drivers: crypto: stm32: h7: support AES GCM and CCM

Adds AES GCM and CCM support to the STM32 crypto driver.

As the implementation has only been validated on a
limited number of STM32 SoCs, it is enabled only
for platforms that set the DT boolean property
`gcm_ccm_supported`.

Note: To use CCM with additional authenticated data (AD),
the heap memory pool size must be large enough to
accommodate the AD length, rounded up to the nearest
multiple of 16 bytes. For an AD length of 8 bytes,
the heap size must be at least 16 bytes. For an AD
length of 17 bytes, you'll need at least 32 bytes.

Signed-off-by: Georgij Černyšiov <geo.cgv@gmail.com>

GeorgeCGV committed:
drivers: crypto: stm32: mark default case as unreachable

Marks default case for cipher mode as unreachable.

Signed-off-by: Georgij Černyšiov <geo.cgv@gmail.com>

mathieuchopstm commented on drivers/crypto/crypto_stm32.c:
```diff
@@ -448,7 +448,7 @@ static int crypto_stm32_session_setup(const struct device *dev,
 			break;
 		default:
-			break;
+			return -EINVAL;
```
`CODE_UNREACHABLE;` would be more sensible, as we already check `mode` earlier in the function:
https://github.com/zephyrproject-rtos/zephyr/blob/6678e29b48e509781563a22b22ec7ce5fced4b93/drivers/crypto/crypto_stm32.c#L377-L382

GeorgeCGV replied:
Ok

mathieuchopstm commented on drivers/crypto/crypto_stm32.c:
```diff
@@ -325,6 +336,109 @@ static int crypto_stm32_ctr_decrypt(struct cipher_ctx *ctx,
+		return ret;
+	}
+
+	if (memcmp(tag, apkt->tag, ctx->mode_params.gcm_info.tag_len) != 0) {
```
Nit: `memcmp` can be/is vulnerable to timing attacks

GeorgeCGV replied:
Discussed in https://github.com/zephyrproject-rtos/zephyr/pull/93189#discussion_r2219571596

mathieuchopstm commented on drivers/crypto/crypto_stm32.c:
```diff
@@ -45,6 +46,12 @@ LOG_MODULE_REGISTER(crypto_stm32);
+#if defined(CONFIG_SOC_STM32H723XX) || defined(CONFIG_SOC_STM32H725XX) ||	\
+	defined(CONFIG_SOC_STM32H730XX) || defined(CONFIG_SOC_STM32H730XXQ) ||	\
+	defined(CONFIG_SOC_STM32H735XX)
+#define STM32_CRYPTO_GCM_CCM_SUPPORTED
```
Nit: define equal to `1` to enable usage of `IS_ENABLED()`/`IF_ENABLED()`.

mathieuchopstm commented on drivers/crypto/crypto_stm32.c:
```diff
@@ -365,21 +479,13 @@ static int crypto_stm32_session_setup(const struct device *dev,
-	if ((mode != CRYPTO_CIPHER_MODE_ECB) &&
-	    (mode != CRYPTO_CIPHER_MODE_CBC) &&
-	    (mode != CRYPTO_CIPHER_MODE_CTR)) {
+#ifndef STM32_CRYPTO_GCM_CCM_SUPPORTED
```
Could be `if (!IS_ENABLED(STM32_CRYPTO_GCM_CCM_SUPPORTED)) {` with patch above

GeorgeCGV replied:
Applied

mathieuchopstm commented on drivers/crypto/crypto_stm32.c:
```diff
@@ -437,6 +439,175 @@ static int crypto_stm32_gcm_decrypt(struct cipher_ctx *ctx, struct cipher_aead_p
+
+	k_sem_give(&data->device_sem);
+
+	k_free(b1);
```
If we're going to always `k_free(b1)`, why not keep it private to `crypto_stm32_ccm`?

GeorgeCGV replied:
Assumed it is used in tag computation, will re-check.

GeorgeCGV replied:
Tested, seems to work fine. Moved to `crypto_stm32_ccm`.

mathieuchopstm commented on drivers/crypto/crypto_stm32.c:
```diff
@@ -437,6 +439,175 @@ static int crypto_stm32_gcm_decrypt(struct cipher_ctx *ctx, struct cipher_aead_p
+
+	k_free(b1);
+
+	if (memcmp(tag, apkt->tag, ctx->mode_params.ccm_info.tag_len) != 0) {
```
Nit: `memcmp` can be/is vulnerable to timing attacks

valeriosetti replied:
>  I don't think Zephyr utils have any constant-time checks.

Yeah, even though the concern is absolutely valid, there are already drivers using `memcmp` for similar task in Zephyr. For example the `crypto_cc23x0.c` one https://github.com/zephyrproject-rtos/zephyr/blob/1f69b91e909cc37301ac15d7e7f30e87c413ab21/drivers/crypto/crypto_cc23x0.c#L878

mathieuchopstm replied:
> Idd. Should I use the working `mbedtls_ct_memcmp` implementation by copying it into the driver? (_Apache 2.0, will need to add extra copyright to mbedlts contributors_). Or leave the check to the upper layer(s)? Some might have mbedtls in their apps. I don't think Zephyr utils have any constant-time checks.

It was more of a general remark than a request. As pointed out by @valeriosetti, another driver also uses `memcmp`.

A simple note in a comment somewhere in the file while keeping usage of `memcmp` is fine by me. Zephyr should provide a constant-time `memcmp` but that's outside PR scope, and can be updated to later.

GeorgeCGV replied:
Considering all of the above, what about adding a Kconfig option to enable/disable the unsafe memcmp function? This would allow users who prefer to implement their tag checks (_e.g., using `mbedtls_ct_memcmp`_) to reduce the driver's footprint.

GeorgeCGV committed:
drivers: crypto: stm32: add mbedtls ct memcmp option

* Use if, endif to scope the configuration options.
* Add an option to use mbedtls_ct_memcmp instead of memcmp.

Signed-off-by: Georgij Černyšiov <geo.cgv@gmail.com>

mathieuchopstm commented on drivers/crypto/crypto_stm32.c:
```diff
@@ -343,18 +653,8 @@ static int crypto_stm32_session_setup(const struct device *dev,
-	 * not a multiple of 128 bits. Therefore, CCM mode is not supported by
-	 * this driver.
-	 */
-	if ((mode != CRYPTO_CIPHER_MODE_ECB) &&
-	    (mode != CRYPTO_CIPHER_MODE_CBC) &&
-	    (mode != CRYPTO_CIPHER_MODE_CTR)) {
```
The old checks should be kept (especially with `CODE_UNREACHABLE` below):
```c
if ((mode != CRYPTO_CIPHER_MODE_ECB) &&
    (mode != CRYPTO_CIPHER_MODE_CBC) &&
    (mode != CRYPTO_CIPHER_MODE_CTR) &&
    (!IS_ENABLED(STM32_CRYPTO_GCM_CCM_SUPPORT) ||
       ((mode != CRYPTO_CIPHER_MODE_CCM)) &&
        (mode != CRYPTO_CIPHER_MODE_GCM))
    )) {
```

GeorgeCGV replied:
Addressed

mathieuchopstm reviewed (commented):
I'm wondering if, instead of hardcoded SoCs in Kconfig, we shouldn't create a new compatible for the IP embedded in H7, as it is also used on other series so the code can probably shared (e.g., STM32F72xx/F73xx).

This can always be integrated later, so OK for me in current state..

erwango reviewed (changes requested):
> I'm wondering if, instead of hardcoded SoCs in Kconfig, we shouldn't create a new compatible for the IP embedded in H7, as it is also used on other series so the code can probably shared (e.g., STM32F72xx/F73xx).

It will be harder to impose to the next ones if this is done this way today. Let's do it now.

erwango commented:
> The Kconfig no longer has dependencies on any particular SoC.

It still does actually. Did you push the right branch ?

GeorgeCGV commented:
> > The Kconfig no longer has dependencies on any particular SoC.
> 
> It still does actually. Did you push the right branch ?

Do you mean? (_.c, not Kconfig_)
```c
#if defined(CONFIG_SOC_STM32H723XX) || defined(CONFIG_SOC_STM32H725XX) ||	\
	defined(CONFIG_SOC_STM32H730XX) || defined(CONFIG_SOC_STM32H730XXQ) ||	\
	defined(CONFIG_SOC_STM32H735XX)
#define STM32_CRYPTO_GCM_CCM_SUPPORT 1
#endif
````

etienne-lms reviewed (commented):
LGTM. Some comments.

GeorgeCGV commented on drivers/crypto/crypto_stm32.c:
```diff
@@ -52,6 +52,14 @@ LOG_MODULE_REGISTER(crypto_stm32);
 #define STM32_CRYPTO_GCM_CCM_SUPPORT 1
 #endif
 
+#if IS_ENABLED(CONFIG_CRYPTO_STM32_USE_MBEDTLS_CT_MEMCMP)
```
Was discussed in https://github.com/zephyrproject-rtos/zephyr/pull/93189#discussion_r2219571596. Ideally, Zephyr shall offer a constant time check function.

Ideally, Zephyr should offer a constant-time check function; it could be added later on in a separate PR.

In the meantime, most developers working with encryption will likely have MbedTLS enabled (_tinycrypt is removed_). Moreover, most drivers don't seem to perform any tag checks.

valeriosetti replied:
> For info, in the long term, [MbedTLS API will no more be supported](https://mbed-tls.readthedocs.io/en/latest/project/long-term-plans/#making-psa-crypto-the-only-crypto-api) so we will sooner or later have to address this.

FWIW what is being referenced there as "legacy interface" only concerns with crypto algorithms; platform support might remain. Currently `mbedtls_ct_memcmp` is still a [public function](https://github.com/Mbed-TLS/TF-PSA-Crypto/blob/development/include/mbedtls/constant_time.h#L32) in TF-PSA-Crypto so, to the best of my knowledge, it will still be available in the next 4.0/1.0 release (expected September this year).

GeorgeCGV replied:
Added your implementation.

valeriosetti replied:
@etienne-lms sorry for the delay, but I've been away for the last 2 weeks.
I checked the current implementation about time constant memcmp() and it looks reasonable to me. A proposal though: currently `crypto_stm32_ct_memcmp()` in only being used in this ST driver, but it seems to me that it can be handy for anyone looking for a similar functionality across the Zephyr codebase. Do you think it might be worth to place it somewhere else so that everybody can make use of it if/when needed?

GeorgeCGV commented on drivers/crypto/crypto_stm32.c:
```diff
@@ -325,6 +338,283 @@ static int crypto_stm32_ctr_decrypt(struct cipher_ctx *ctx,
+	}
+
+	/* memcmp is vulnerable to timing attacks */
+	if (memcmp(tag, apkt->tag, ctx->mode_params.gcm_info.tag_len) != 0) {
```
The `tag` is local; the executing `apkt` and `ctx` are not expected to change. The semaphore is there to control hardware access, which plays no role in tag comparison; or am I missing something?

GeorgeCGV committed:
drivers: crypto: stm32: add constant time comparison

Adds constant time comparison to replace memcmp.

Co-Authored-By: Etienne Carriere <etienne.carriere@st.com>
Signed-off-by: Georgij Černyšiov <geo.cgv@gmail.com>

GeorgeCGV committed:
dts: bindings: crypto: stm32: add gcm-ccm-supported

Adds a gcm-ccm-supported property to the crypto driver.
This allows to enable or disable support for AES GCM and CCM,
providing flexibility to remove the functionality
if it's not needed.

Signed-off-by: Georgij Černyšiov <geo.cgv@gmail.com>

GeorgeCGV committed:
dts: arm: st: stm32h730: enable aes gcm ccm support

Enables AES GCM and CCM support for the crypto driver.

Signed-off-by: Georgij Černyšiov <geo.cgv@gmail.com>

GeorgeCGV commented:
@etienne-lms added your constant time check implementation with a minor modification to return an int.
@erwango added gcm ccm support property as discussed on Discord.

Rebased.

GeorgeCGV commented on drivers/crypto/crypto_stm32.c:
```diff
@@ -81,7 +84,13 @@ typedef status_t (*hal_cryp_aes_op_func_t)(CRYP_HandleTypeDef *hcryp, uint8_t *i
+#define hal_gcm_encrypt_op hal_encrypt
+#define hal_gcm_decrypt_op hal_decrypt
+#define hal_ccm_encrypt_op hal_encrypt
+#define hal_ccm_decrypt_op hal_decrypt
```
not mandatory, opted for consistency; changed

erwango reviewed (commented):
LGTM, but I have a last request. Would you min enabling aes node on `stm32h735g_disco` so this part of code could be compiled on crypto sample in CI ?
Maybe you could just replace the existing STM32 integration board as this one will cover more code than existing one.

erwango commented:
> STM32H723XX, STM32H725XX, STM32H730XX, STM32H730XXQ, and STM32H735XX, as they share the same reference manual.

According to HAL, this is rather actually supported on STM32H730/3/5, STM32H750/3/5/7 and STM32H7B0/3.
Would you mind to add the new prop to H755 and H7B0 dtsi files?
 
Then it seems we'll need some clean un in these files as `cryp@48021000` node in H755 should actually be moved to H750, but this is outside the scope of this PR.

GeorgeCGV committed:
boards: stm32h735g_disco: enable crypto

Enables crypto peripheral for the board.

Signed-off-by: Georgij Černyšiov <geo.cgv@gmail.com>

GeorgeCGV committed:
dts: arm: st: stm32h755: enable aes gcm ccm support

Enables AES GCM and CCM support for the crypto driver.

Signed-off-by: Georgij Černyšiov <geo.cgv@gmail.com>

GeorgeCGV committed:
dts: arm: st: stm32h7b0: enable aes gcm ccm support

Enables AES GCM and CCM support for the crypto driver.

Signed-off-by: Georgij Černyšiov <geo.cgv@gmail.com>

GeorgeCGV commented:
>  Would you mind to add the new prop to H755 and H7B0 dtsi files?

I didn't test on them; therefore, I don't know if it works there. Added.

erwango reviewed (dismissed):
Commit message of "drivers: crypto: stm32: h7: support AES GCM and CCM" would require an update for a copy/paste issue and information likely not correct.

GeorgeCGV committed:
soc: st: stm32: set minimal heap when crypto with ccm is enabled

The AES CCM AD (Additional Authenticated Data) relies on
dynamic memory allocation, the heap memory pool size
must be greater than zero.

This commit sets the heap memory one block size when you
enable crypto with CCM support. This ensures that the crypto
sample can be built and run.

Signed-off-by: Georgij Černyšiov <geo.cgv@gmail.com>

erwango reviewed (changes requested):
Otherwise LGTM

etienne-lms reviewed (commented):
Thanks for the update.

GeorgeCGV commented:
Addressed recent review comments and corrected padded buffer allocation for AES CCM (_there was an issue with overpadding_).

GeorgeCGV commented on soc/st/stm32/Kconfig.defconfig:
```diff
@@ -111,4 +111,14 @@ config VIDEO_BUFFER_POOL_ALIGN
+
+# Set the minimum required heap size to enable
+# AES CCM with Additional Data (AD) support
+# for running samples/drivers/crypto sample.
```
Modified the comment to mention how many AD data bytes are supported, as well as to remove the sample mention.

The goal is to enable crypto to be built without the heap warning on targets that support CCM and GCM. The warning is necessary to prevent potential issues when the heap is not present (_default case_).

etienne-lms reviewed (dismissed):
Thanks @GeorgeCGV for your patience and this driver.
LAGTM.

etienne-lms reviewed (dismissed):
Sorry for the late feedback. Looks all good to me.

GeorgeCGV commented:
Rebased

etienne-lms commented:
Keep alive.
We'll soon update hal_stm32 module for H7 that will address https://github.com/STMicroelectronics/stm32h7xx-hal-driver/issues/88.
Note however that  the CCM issue https://github.com/STMicroelectronics/stm32h7xx-hal-driver/issues/104 will not yet be addressed.

etienne-lms commented on dts/bindings/crypto/st,stm32-cryp.yaml:
```diff
@@ -6,3 +6,8 @@ description: STM32 Cryptographic Accelerator
+properties:
+  gcm-ccm-supported:
+    type: boolean
+    description: Enables AES GCM and CCM support.
```
All "st,stm32-cryp" compatible peripherals actually do support GCM and CCM. I think this extra DT property could be removed. That said, I'l fine you currently keep it and we'll remove it later once integration on all applicable SoCs is explicitly tested against GCM/CMM support.

GeorgeCGV replied:
The idea was to exclude the not-tested MCUs

etienne-lms replied:
Ok, fine as-is.

GeorgeCGV commented:
basic rebase. Some unrelated tests seem to fail, will rebase later on

GeorgeCGV commented:
@etienne-lms tested CCM part against https://github.com/STMicroelectronics/stm32h7xx-hal-driver/commit/cd566efa4e52854f5780e9c022a12af5bda1ad78, the commit resolves overflow and byte input to the HW. Are there plans to update https://github.com/zephyrproject-rtos/hal_stm32 in the near future to grab that h7 commit?

erwango commented:
> @etienne-lms tested CCM part against [STMicroelectronics/stm32h7xx-hal-driver@cd566ef](https://github.com/STMicroelectronics/stm32h7xx-hal-driver/commit/cd566efa4e52854f5780e9c022a12af5bda1ad78), the commit resolves overflow and byte input to the HW. Are there plans to update https://github.com/zephyrproject-rtos/hal_stm32 in the near future to grab that h7 commit?

Did you check it isn't part of quite fresh https://github.com/zephyrproject-rtos/zephyr/pull/111050 ?

GeorgeCGV commented:
@erwango hi, https://github.com/zephyrproject-rtos/zephyr/pull/111050 mentions [STM32H7 v1.13.0](https://github.com/STMicroelectronics/STM32CubeH7/tree/v1.13.0/Drivers) that uses [stm32h7xx-hal-driver release v1.11.6](https://github.com/STMicroelectronics/stm32h7xx-hal-driver/commit/a1996eed9172b59887bafaaa0ea1816ea14d48b5) which doesn't include:

- `[HAL][SMARTCARD] Update configuration sequence to better fit SMARTCARD specifications`
- `[HAL][CRYP] Fix CCM partial-word output handling for Non-Multiple of 4 Byte Payloads` - addresses ccm issues
- `[HAL][DMA] Fixed incorrect callback execution due to inverted CT bit check in HAL_DMA_IRQHandler` 
- `[LL][GENERIC] Add new LL APIs to configure the VREFBUF peripheral` - addresses #102301

erwango commented:
> @erwango hi, #111050 mentions [STM32H7 v1.13.0](https://github.com/STMicroelectronics/STM32CubeH7/tree/v1.13.0/Drivers) that uses [stm32h7xx-hal-driver release v1.11.6](https://github.com/STMicroelectronics/stm32h7xx-hal-][...]

Feel free to push a cherry-pick PR on hal_stm32.

GeorgeCGV commented:
Created, rebased

GeorgeCGV commented:
Added an option to workaround tag corruption during AES CCM decryption, mentioned in https://github.com/STMicroelectronics/stm32h7xx-hal-driver/issues/104#issuecomment-4778739244.

etienne-lms commented on drivers/crypto/crypto_stm32.c:
```diff
@@ -53,17 +53,30 @@ LOG_MODULE_REGISTER(crypto_stm32);
+		diff |= *pa ^ *pb;
+	}
+
+	return (int) diff;
```
Nitpicking (and non-blocking); prefer no space char between the cast operator and the cast'ed value.
```suggestion
	return (int)diff;

GeorgeCGV replied:
corrected

etienne-lms commented on drivers/crypto/crypto_stm32.c:
```diff
@@ -614,6 +623,10 @@ static int crypto_stm32_ccm_decrypt(struct cipher_ctx *ctx, struct cipher_aead_p
 
+#ifdef CONFIG_CRYPTO_STM32_AES_CCM_TRAILING_BYTES_FIXUP
+	memcpy(apkt->tag, apkt_tag, ctx->mode_params.ccm_info.tag_len);
+#endif
```
Should `apkt->tag` input content be preserved even upon decryption failure?

GeorgeCGV replied:
Unknown, I couldn't find guidance on how the crypto layer shall handle buffers and tags on failure (e.g., whether the output buffer should be zeroed on error).

GeorgeCGV replied:
Additionally, the driver doesn't validate that all input and output lengths are correct.
When integrated into PSA/MbedTLS - that layer provides all the checks. When used outside, then no checks are there. But I don't see a lot of defensive checks across other crypto drivers.

GeorgeCGV replied:
But considering https://github.com/zephyrproject-rtos/zephyr/issues/43712, it might not be worth putting much effort into the crypto drivers that are expected to be deprecated.

GeorgeCGV replied:
Added additional checks as described in https://github.com/zephyrproject-rtos/zephyr/pull/93189#issuecomment-4817929764, but zeroization of buffers upon failure is not a part of the PR (_has to be decided elsewhere and applied across all crypto drivers if required_).

etienne-lms commented on drivers/crypto/crypto_stm32.c:
```diff
@@ -604,6 +604,15 @@ static int crypto_stm32_ccm_decrypt(struct cipher_ctx *ctx, struct cipher_aead_p
+	if (ctx->mode_params.ccm_info.tag_len > sizeof(apkt_tag)) {
+		return -EINVAL;
+	}
+	memcpy(apkt_tag, apkt->tag, ctx->mode_params.ccm_info.tag_len);
```
This looks a bit fragile. Maybe it's not the tag that is located right after the input decryption buffer, not some other data that would get corrupted. This workaround looks valid but only if `apkt->pkt->in_buf + apkt->pkt->in_len == apkt->tag`. In other cases, if `apkt->pkt->in_len % 4 != 0`, the driver should error out. At least, we know the tag is >= 4 bytes.

GeorgeCGV replied:
True, users can provide a tag that can be located elsewhere - in that case, the HAL crypto function will zero out some bytes beyond the output buffer if the buffer length is not a multiple of 4. That might lead to crashes. However, if the buffer length is fine but the payload isn't, then nothing bad will happen. 

Do you suggest adding a check that the tag follows the `out` buffer? And fail if it is not the case, but the workaround is enabled? Or fail all the time when such a condition is true, but then the decryption will fail for some valid cases (_i.e., when the `out` buffer length is correct_)?

etienne-lms replied:
I think unless we're sure HAL won't scratch uncontrolled data, we should refuse to proceed. We can't know the real size allocated for the buffer, only the payload size so we have to deal with it.

A way to overcome your issues would be a possibly costly memory allocation of the input buffer, or by smaller chunks, anyhow at the price of performances due required memory copy. Maybe consider decrypting aligned data first and last unaligned 3 bytes using a temp input buffer. Or we find a way to fix the HAL...

GeorgeCGV replied:
I see only a couple of ways: without HAL modification (_preferred way_):

- heap-allocated aligned tmp buff - allocation + memcpy, probably the safest, but also the slowest.
- ram allocated aligned tmp buff of a size defined via Kconfig - easy to implement, but unnecessary data duplication, user has to fine-tune the buffer size (_removes allocation_)
- in place fix by persistence and restoration of the expected overflow bytes - more performant, risky, but can work (_removes allocation, reduces memcpy to [1,3] bytes_).
- HAL fix - would be the preferred way, as the data can be filtered/modified during the process cycle (_also fixes the issue for non-Zephyr users_). I'm not sure why it was necessary to [zero-fill trailing bytes](https://github.com/STMicroelectronics/stm32h7xx-hal-driver/blob/7a9433030448812bf845dd38ea6c8e71e609c565/Src/stm32h7xx_hal_cryp.c#L2933) in the destination buffer. Zeroization of trailing bytes (_padding bytes_) is most probably a mistake, as it touches outside memory for no reason. 

PS: as HAL uses `CRYP_CopyPartialOutputWord` for CCM and GCM, both ciphers are affected.

etienne-lms replied:
> in place fix by persistence and restoration of the expected overflow bytes - more performant, risky, but can work (removes allocation, reduces memcpy to [1,3] bytes).

I don't think it's the right way. Maybe the modified data is being used while the crypto operation is processing.

I think the HAL API allows multi-stage operations: first process the input buffer from start to last 4-byte aligned offset sample, then process the 1 to 3 remaining bytes (if any) using a temp buffer from the stack, and finally complete the operation for the authentication part. Would that make sense to you?

Note however that Zephyr test harness on crypto operation is quite weak. To check nothing unexpected happens, a stronger test suite should rather be used, that's maybe off-topic, it's rather a wide Zephyr issue, worked around when using MbedTLS/PSA-Crypto that is tested internally (I mean outside the Zephyr application scope).

> I'm not sure why it was necessary to [zero-fill trailing bytes](https://github.com/STMicroelectronics/stm32h7xx-hal-driver/blob/7a9433030448812bf845dd38ea6c8e71e609c565/Src/stm32h7xx_hal_cryp.c#L2933) in the destination buffer. Zeroization of trailing bytes (padding bytes) is most probably a mistake, as it touches outside memory for no reason.

I would tend to agree with you, but i'd need a deeper look into the refman to understand.

GeorgeCGV replied:
>  think the HAL API allows multi-stage operations: first process the input buffer from start to last 4-byte aligned offset sample, then process the 1 to 3 remaining bytes (if any) using a temp buffer from the stack, and finally complete the operation for the authentication part. Would that make sense to you?

Confirmed feasible. But to support reusable crypto sessions, the driver needs a minor state adjustment to the peripheral state to skip the IV and Key configurations during sequential encryption blocks.

I still have fail cases after running some tests; will see.

GeorgeCGV replied:
@etienne-lms modified ccm, passes all tests so far (_block-sized and less than block-sized encryption and decryption payloads for all possible paddings_).

Main change: CCM is now processed in blocks, with each block allocated on the stack. The output is first written to a staging buffer and then copied to the `out` buffer.

The same approach will likely need to be applied to GCM, but this still needs to be tested.

PS: used `hal_stm32` updated to https://github.com/STMicroelectronics/stm32h7xx-hal-driver/commit/cd566efa4e52854f5780e9c022a12af5bda1ad78.

GeorgeCGV replied:
Adapted GCM as well, now it passes the tests too.

@etienne-lms the ztests can be found here:

- [Crypto AES CCM roundtrip tests for overflow/zeroization detection](https://gist.github.com/GeorgeCGV/5c234da934024f132398ab413defafb8)
- [Crypto AES GCM roundtrip tests for overflow/zeroization detection](https://gist.github.com/GeorgeCGV/e83ec3725d009176dca3dfd269478a13)

Unfortunately, H723zg doesn't have crypto peripheral and I don't have any other in-tree board (_out of supported in the PR_) to integrate (_add, run_) the tests into the upstream.

GeorgeCGV replied:
Re-tested with https://github.com/zephyrproject-rtos/hal_stm32/commit/ff19fb0b03bc9b3f1355e9f9b7b6b3a6c051f121, works too (_without the https://github.com/STMicroelectronics/stm32h7xx-hal-driver/commit/cd566efa4e52854f5780e9c022a12af5bda1ad78_); therefore, removing `DNM` label.

Potential improvements (_outside of the PR_):
* CCM: AD buffer can be pre-allocated in RAM, with max size configured via Kconfig. Avoids heap, slightly faster.
* Peripheral de-init and re-init could be disabled, i.e. init it once and leave. Slightly better performance for some use cases.

etienne-lms replied:
This is addressed now.

etienne-lms commented on drivers/crypto/crypto_stm32.c:
```diff
@@ -325,6 +337,366 @@ static int crypto_stm32_ctr_decrypt(struct cipher_ctx *ctx,
+		session->config.HeaderWidthUnit = CRYP_HEADERWIDTHUNIT_BYTE;
+	}
+
+	/* run operation in full blocks, then the remaining bytes */
```
I think an explanation would be nice here. Something like `STM32 HAL for GCM expects input data are 4byte aligned hence if not, use a local bounce buffer to process the last samples.`

Ditto in `crypto_stm32_ccm()`

GeorgeCGV replied:
Added more clarification. It is not only 4-byte alignment, but data processing shall be done in blocks to prevent zero-pad bytes inserted by the HAL (_for payloads > N blocks + trailing X bytes_).

Commit message updated.

etienne-lms reviewed (commented):
The integration looks good to me. A good think you ensure we don't trigger the ugly access implemented in the HAL. I hope it will be addressed, one way or another.

I'll go again through your changes but they look good to me.

Note in commit *"drivers: crypto: stm32: h7: support AES GCM and CCM"*:
the commit message mentions H723XX, * H725XX, but as you said, these do not embed a CRYP block. Moreover the commit (and the later changes) add this support for all H7 SoCs embedding a CRYP instance. Could you rephrase the commit message to something like `Adds AES GCM and CCM support in STM32 crypto drivers. The implementation has been validated only upon a small amount of STM32 SoCs hence is enabled only upon DT boolean property 'gcm_ccm_supported'`?

We will later test on other STM32 applicable SoCs and extend `gcm_ccm_supported` where applicable in later changes. IMHO, I'll end up removing this property as all CRYP and AES blocks in STM32 SoCs are expected to support GCM/CCM. Yet HAL this driver relies on is the be checked :-)

GeorgeCGV commented:
@etienne-lms rebased and some updates:

- Factored out staged operation.
- Added:
  - Zero-payload handling
  - `NULL` checks where applicable
  - Checks for output and input buffer length
  - Support for in-place operation (_when `apkt->pkt->out_buf` is `NULL`_)
  - AD length overflow check

Updated ztests gists ([Crypto CCM tests](https://gist.github.com/GeorgeCGV/5c234da934024f132398ab413defafb8), [Crypto GCM tests](https://gist.github.com/GeorgeCGV/e83ec3725d009176dca3dfd269478a13)):
- Added in-place round-trip test
- Added empty payload test

All tests pass on H730.

etienne-lms commented on drivers/crypto/crypto_stm32.c:
```diff
@@ -283,26 +333,413 @@ static int crypto_stm32_ctr_encrypt(struct cipher_ctx *ctx,
+		if (ret < 0) {
+			return ret;
+		}
+		return 0;
```
Nitpicking:
```suggestion
		return do_aes(ctx, fn, staging_in, 0, staging_out);

etienne-lms commented on drivers/crypto/crypto_stm32.c:
```diff
@@ -283,26 +333,413 @@ static int crypto_stm32_ctr_encrypt(struct cipher_ctx *ctx,
+	uint8_t *dst;
+	uint32_t iv[BLOCK_LEN_WORDS] = {0};
+
+	if ((apkt->pkt->in_len < 0) || (apkt->pkt->out_buf_max < apkt->pkt->in_len)) {
```
```suggestion
	if (!IN_RANGE(apkt->pkt->in_len, 0, apkt->pkt->out_buf_max) {

GeorgeCGV replied:
Thanks, adapted

etienne-lms reviewed (commented):
LGTM. 2 minor non-blocking suggestions.

erwango commented:
@valeriosetti PTAL
